clustermesh-apiserver/clustermesh/synchronizer.go

```diff
@@ -66,45 +84,136 @@ func RegisterSynchronizer[T runtime.Object](in syncParams[T]) {
+				} else {
+					logger.Debug("Namespace watcher is not enabled for resource type")
+				}
 
-					process := func(op, key string, do func() error) {
-						logger.Info("Updating resource in etcd",
-							logfields.Operation, op,
-							logfields.Key, key,
-						)
-
-						if err := do(); err != nil {
-							logger.Warn("Failed updating resource in etcd",
-								logfields.Error, err,
-								logfields.Operation, op,
-								logfields.Key, key,
-							)
+				for {
+					select {
+					case event, ok := <-resourceEvents:
+						if !ok {
+							logger.Debug("Resource event channel closed, stopping synchronizer")
+							resourceEvents = nil
+							continue
 						}
-					}
 
-					upserts, deletes := in.Converter.Convert(event)
-					for upsert := range upserts {
-						process("upsert", upsert.GetKeyName(), func() error { return store.UpsertKey(ctx, upsert) })
-					}
-					for delete := range deletes {
-						process("delete", delete.GetKeyName(), func() error { return store.DeleteKey(ctx, delete) })
+						if event.Kind == resource.Sync {
+							event.Done(nil)
+							logger.Info("Initial entries successfully received from Kubernetes")
+							store.Synced(ctx, synced)
+							continue
+						}
+						// Filter the event based on namespace global status.
+						// Only required for certain resource types.
+						// Ignore delete events as they should always be processed.
+						if event.Kind != resource.Delete && in.Options.Namespaced {
+							ns := in.Namespacer.ExtractNamespace(event)
+							// Check if the namespace is empty.
+							if ns == "" {
+								logger.Error("Failed to determine namespace for resource event, skipping",
+									logfields.Name, event.Key.Name,
+								)
+								// No way to process this event, just mark done and continue.
+								event.Done(nil)
+								continue
+							}
+							isGlobal, err := in.NamespaceManager.IsGlobalNamespaceByName(ns)
+							if err != nil {
+								logger.Warn("Failed to determine if namespace is global",
+									logfields.Error, err,
+									logfields.K8sNamespace, ns,
+								)
+								// Retry this as it might succeed later because of dependency on namespace store.
+								event.Done(err)
+								continue
+							}
+							if !isGlobal {
+								logger.Debug("Skipping resource event as it is not in a global namespace",
+									logfields.Name, event.Key.Name,
+									logfields.K8sNamespace, ns,
+								)
+								// There is a possible edge case here.
+								// For resources like CiliumEndpointSlice, where namespace is
+								// inferred from mutable field, it is possible that the resource
+								// was previously in a global namespace and is now moved to a
+								// non-global namespace. In this case, we should ensure that
+								// the resource is deleted from kvstore. Mutate the event to
+								// be a delete and continue processing. event.Done will be called
+								// later.
+								event.Kind = resource.Delete
```

Comment:
I am not sure I understand the comment here, I reckon that CiliumEndpointSlice might be an edge case but it doesn't seems to be specifically handled here? Making it a delete event is the handling for everything so I am a bit confused, the comment might just need to be changed a bit :man_shrugging: